### DISCUSSION TITLE
Is there a way to create a test server for minimal APIs without WebApplicationFactory<TEntryPoint>?

### DISCUSSION THREAD

**Discussion by paulomorgado** [dotnet/aspnetcore#51782] [Q&A]
I'd like to have something like this on my tests:

```csharp
var messages = new List<MyMessage>();

var builder = WebApplication.CreateBuilder(args);

var app = builder.Build();

app.MapGet("/", (MyMessage m) => messages.Add(m));

app.Run();
```

Now I would like to have a `TestServer` instance for that and use:

```csharp
await testServer.CreateClient().PostAsJsonAsync(new MyMessage);
// check messages
```

But this seems to be possible only using `WebApplicationFactory<TEntryPoint>`. And that requires an entrypoint.

**Comment by martincostello** [comment]
You need to do [something like this](https://github.com/martincostello/dotnet-minimal-api-integration-testing/blob/4cf758cb4a7041c3ef729cf40494d00891623f58/src/TodoApp/Program.cs#L71-L74) to expose the `Program` class, then you can use it [like this](https://github.com/martincostello/dotnet-minimal-api-integration-testing/blob/4cf758cb4a7041c3ef729cf40494d00891623f58/tests/TodoApp.Tests/TodoAppFixture.cs#L19).

### Program.cs

```csharp
var messages = new List<MyMessage>();

var builder = WebApplication.CreateBuilder(args);

var app = builder.Build();

app.MapGet("/", (MyMessage m) => messages.Add(m));

app.Run();

public partial class Program
{
    // Expose the Program class for use with WebApplicationFactory<T>
}
```

### Your Tests

```csharp
var factory = new WebApplicationFactory<Program>();
```

**Reply by paulomorgado** [reply]
That still requires `WebApplicationFactory<TEntryPoint>`.

What if I need to setup 2 or 3 hosts for my test? Does it support multiples setups in the same assembly?

**Reply by martincostello** [reply]
As far as I'm aware, no, that is not supported by `WebApplicationFactory<T>`. If you need more than one entry point exposed in public API, you need to not use top-level programs.

**Reply by paulomorgado** [reply]
Besides that, `messages` is supposed to be an instance supplied for each server created.

**Comment by paulomorgado** [comment]
Just checked. `TEntryPoint` must be on an assembly with an entry point that builds a host.

If it's a unit tests project, that doesn't work.
